src/lib/tls/tls13_pqc/hybrid_public_key.cpp

```diff
@@ -0,0 +1,373 @@
+      m_public_keys, size_t(0), [](size_t es, const auto& key) { return std::max(es, key->estimated_strength()); });
+}
+
+size_t Hybrid_KEM_PublicKey::key_length() const {
+   return reduce(m_public_keys, size_t(0), [](size_t kl, const auto& key) { return kl + key->key_length(); });
+}
+
+bool Hybrid_KEM_PublicKey::check_key(RandomNumberGenerator& rng, bool strong) const {
+   return reduce(m_public_keys, true, [&](bool ckr, const auto& key) { return ckr && key->check_key(rng, strong); });
+}
+
+AlgorithmIdentifier Hybrid_KEM_PublicKey::algorithm_identifier() const {
+   throw Botan::Not_Implemented("Hybrid keys don't have an algorithm identifier");
+}
+
+std::vector<uint8_t> Hybrid_KEM_PublicKey::public_key_bits() const {
+   // Technically, that's not really correct. The docstring for public_key_bits()
+   // states that it should return a BER-encoding of the public key.
+   //
+   // TODO: Perhaps add something like Public_Key::raw_public_key_bits() to
+   //       better reflect what we need here.
+   return public_value();
+}
+
+std::vector<uint8_t> Hybrid_KEM_PublicKey::public_value() const {
+   // draft-ietf-tls-hybrid-design-06 3.2
+   //   The values are directly concatenated, without any additional encoding
+   //   or length fields; this assumes that the representation and length of
+   //   elements is fixed once the algorithm is fixed.  If concatenation were
+   //   to be used with values that are not fixed-length, a length prefix or
+   //   other unambiguous encoding must be used to ensure that the composition
+   //   of the two values is injective.
+   return reduce(m_public_keys, std::vector<uint8_t>(), [](auto pkb, const auto& key) {
+      // Technically, this is not correct! `public_key_bits()` is meant to
+      // return a BER-encoded public key. For (e.g.) Kyber, that contract is
+      // broken: It returns the raw encoding as used in the reference
+      // implementation.
+      //
+      // TODO: Provide something like Public_Key::raw_public_key_bits() to
+      //       reflect that difference. Also: Key agreement keys could return
+      //       their raw public value there.
+      return concat(pkb, key->public_key_bits());
+   });
+}
+
+bool Hybrid_KEM_PublicKey::supports_operation(PublicKeyOperation op) const {
+   return PublicKeyOperation::KeyEncapsulation == op;
+}
+
+namespace {
+
+class Hybrid_KEM_Encryption_Operation final : public PK_Ops::KEM_Encryption_with_KDF {
+   public:
+      Hybrid_KEM_Encryption_Operation(const Hybrid_KEM_PublicKey& key,
+                                      std::string_view kdf,
+                                      std::string_view provider) :
+            PK_Ops::KEM_Encryption_with_KDF(kdf) {
+         std::transform(
+            key.public_keys().begin(),
+            key.public_keys().end(),
+            std::back_inserter(m_kem_encryptors),
+            [&](const auto& pubkey) { return std::make_unique<PK_KEM_Encryptor>(*pubkey, "Raw", provider); });
```

Comment:
I know above is considered more modern C++ style (I guess) but why not

```
for(const auto& k: key.public_keys()) {
  m_kem_encryptors.push_back(std::make_unique<PK_KEM_Encryptor>(*k, "Raw", provider));
}
```

Should have identical performance characteristics, and 3 fewer lines with less repetition.

Reply:
@FAlbertDev will have a good laugh about this. 🤣 Yeah! I agree. Looking forward to when we can leverage ranges. (I believe macOS and NDK weren't yet there, when I last tried)